### ISSUE TITLE
Make Agda smarter, avoid some "the solution cannot be used at relevant, unrestricted modality" errors?

### ISSUE THREAD

**Issue by nad** [agda/agda#4366]
This issue has been broken out of #4174.

Consider the following old code snippet, from a development by @UlfNorell and myself:
```agda
mutual

  data Nonterminal : NonTerminalType where
    e : Nonterminal _ Char

  grammar : Grammar Nonterminal Char
  grammar e = tok '0' ⊛> tok '+' ⊛> ! e
            ∣ tok '0'
```
After recent changes (see #4174) the underscore is not solved "because the solution cannot be used at relevant, unrestricted modality". This can be fixed in the following way:
```agda
mutual

  private

    i = _

  data Nonterminal : NonTerminalType where
    e : Nonterminal i Char

  grammar : Grammar Nonterminal Char
  grammar e = tok '0' ⊛> tok '+' ⊛> ! e
            ∣ tok '0'
```
However, I think this is a bit ugly. Couldn't Agda be a little smarter?

@UlfNorell wrote the following:

> If you have a metavariable that could be instantiated with an erased term, we should refuse to solve a non-erased argument to a term containing this meta.

If the meta-variable is later instantiated to a term that does not contain any erased subterms, then this solution is fine. Could we postpone the solving until the meta-variable has been fully instantiated?

**Comment by guillaumebrunerie** [comment]
Shouldn’t this be labelled as a regression instead?
Some of my code broke after updating my Agda due to this issue (and your workaround worked fine, except that I have nine constructors in my datatype, therefore I need nine "i = _"…).

Here is a self-contained example that I wrote before noticing that the issue had already been reported:

```agda
postulate
  A : Set
  P : A → Set

  a : A
  p : P a

record R : Set₁ where
  field
    S : A → Set
    f : {x : A} (r : S x) → P x
open R public


-- We define [r] and [D] mutually so that we do not have to give the implicit
-- argument in the type of [d]
r : R

data D : {x : A} → Set where
  d : D

S r x = D {x}
f r {x} d = p  -- works on 2.6.0.1, fails on master
```

**Comment by UlfNorell** [comment]
In #4442 [Jesper wrote](https://github.com/agda/agda/issues/4442#issuecomment-586174574)

> Maybe one way to fix it would be to instead 'promote' the erased meta by creating a new meta of the same type that is not erased, and instantiating the erased one with it.

**Comment by andreasabel** [comment]
Milestone 2.6.1 transferred from duplicate issue #4442.

**Comment by jespercockx** [comment]
> In #4442 [Jesper wrote](https://github.com/agda/agda/issues/4442#issuecomment-586174574)
> 
> > Maybe one way to fix it would be to instead 'promote' the erased meta by creating a new meta of the same type that is not erased, and instantiating the erased one with it.

I am retracting my words: I thought the problem was with the unifier for constraint solver, but actually it is a problem with the pattern matching unifier. However, the pattern matching unifier is not supposed to instantiate any metavariables, so we cannot use the trick I was thinking of.

> If the meta-variable is later instantiated to a term that does not contain any erased subterms, then this solution is fine. Could we postpone the solving until the meta-variable has been fully instantiated?

We already do postpone the checking of the left-hand side when it is blocked on a metavariable. The problem (at least in the example by @guillaumebrunerie) is that the metavariable only gets instantiated when checking the body of the second clause of `p`, but the typechecker never gets to that point because it is stuck on checking the LHS of that clause.

**Comment by jespercockx** [comment]
@guillaumebrunerie your example can easily be fixed by writing `d : D {a}` instead of `d : D`, there's no need for additional `i = _`s.

**Comment by jespercockx** [comment]
I cannot think of a way to fix this problem without introducing a lot of additional complexity. I also don't think the example by @guillaumebrunerie is particularly motivating, since it relies on a metavariable in the type of a constructor being solved only while checking a definition that pattern matches on this datatype, which seems wrong. I think it would anyway be better to just write your datatypes in a way that they make sense on their own, without having to rely on other definitions to infer the types of the constructors.

**Comment by guillaumebrunerie** [comment]
Here is an unobfuscated code showing where/how I use this pattern.
But it does seem a little bit like a hack anyway, so I’m fine with it not being fixed.

```agda
postulate
  -- [Arity] represents a type of arities
  Arity : Set
  -- [Op a] represents the type of all operations of arity [a]
  Op : Arity → Set

  -- Some examples of arities and corresponding operations
  a1 a2 a3 a4 : Arity
  Op1 : Op a1
  Op2 : Op a2
  Op3 : Op a3
  Op4 : Op a4

-- A signature is a type indexed by arities. If [Σ] is of type [Signature], then
-- [Σ {a}] represents the type of operations of arity [a] in the signature [Σ]
Signature : Set₁
Signature = {a : Arity} → Set

-- A model is a signature together with operations of the correct arity for
-- everything in the signature
record Model : Set₁ where
  field
    sig : Signature
    ops : {a : Arity} → sig {a} → Op a
open Model


-- Here is now a concrete example of a model. The arities of [op1]/… are
-- determined by the types of [Op1]/…, so it is nice if Agda can figure them out.

M : Model

data Σ : Signature where
  op1 : Σ
  op2 : Σ
  op3 : Σ
  op4 : Σ

sig M {a} = Σ {a}
ops M op1 = Op1
ops M op2 = Op2
ops M op3 = Op3
ops M op4 = Op4
```

**Comment by nad** [comment]
> it relies on a metavariable in the type of a constructor being solved only while checking a definition that pattern matches on this datatype, which seems wrong.

Why is this wrong, if you specifically ask for it?

I do not think that we should introduce additional complexity only for this, seemingly rarely used, design pattern.

**Comment by jespercockx** [comment]
Maybe 'wrong' is a too strong word, but I would think it is better for readability of the code if datatype definition can be made sense of on their own.

**Comment by nad** [comment]
In the case of the code mentioned in the OP I did not want to read or write the index, and the index could be uniquely inferred from the function. Giving the same information twice might make the code less readable.

**Comment by jespercockx** [comment]
> I do not think that we should introduce additional complexity only for this, seemingly rarely used, design pattern.

Well, I had never seen it before, but apparently there are at least two (very intensive) users of Agda who use it.
